File: core/src/main/java/ru/deadsoftware/cavedroid/menu/MenuScreen.java
package ru.deadsoftware.cavedroid.menu;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.Screen;
import ru.deadsoftware.cavedroid.MainConfig;
import ru.fredboy.cavedroid.domain.menu.repository.MenuButtonRepository;
import ru.fredboy.cavedroid.zygote.menu.input.MenuInputProcessor;
import ru.fredboy.cavedroid.zygote.menu.renderer.MenuRenderer;

import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
public class MenuScreen implements Screen {

    private final MenuRenderer mMenuRenderer;

    private final MenuInputProcessor mMenuInputProcessor;

    private final MenuButtonRepository mMenuButtonsRepository;

    @Inject
    public MenuScreen(MainConfig mainConfig) {
        MenuComponent menuComponent = DaggerMenuComponent.builder()
                .mainComponent(mainConfig.getMainComponent()).build();

        mMenuRenderer = menuComponent.menuRenderer();
        mMenuInputProcessor = menuComponent.menuInputProcessor();
        mMenuButtonsRepository = menuComponent.menuButtonsRepository();
    }

    public void resetMenu() {
        mMenuButtonsRepository.setCurrentMenu("main");
    }

    @Override
    public void show() {
        Gdx.input.setInputProcessor(mMenuInputProcessor);
    }

    @Override
    public void render(float delta) {
        mMenuRenderer.render(delta);
    }

    @Override
    public void resize(int width, int height) {

    }

    @Override
    public void pause() {

    }

    @Override
    public void resume() {

    }

    @Override
    public void hide() {

    }

    @Override
    public void dispose() {

    }
}

File: core/zygote/src/main/kotlin/ru/fredboy/cavedroid/zygote/menu/MenuScreen.kt
package ru.fredboy.cavedroid.zygote.menu

import com.badlogic.gdx.Gdx
import ru.fredboy.cavedroid.domain.configuration.repository.ApplicationContextRepository
import ru.fredboy.cavedroid.domain.menu.repository.MenuButtonRepository
import ru.fredboy.cavedroid.zygote.CaveDroidApplication
import ru.fredboy.cavedroid.zygote.base.BaseScreen
import ru.fredboy.cavedroid.zygote.menu.di.DaggerMenuComponent
import ru.fredboy.cavedroid.zygote.menu.di.MenuComponent
import ru.fredboy.cavedroid.zygote.menu.input.MenuInputProcessor
import ru.fredboy.cavedroid.zygote.menu.renderer.MenuRenderer
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class MenuScreen @Inject constructor(
    applicationContextRepository: ApplicationContextRepository,
) : BaseScreen(applicationContextRepository) {

    private val menuRenderer: MenuRenderer

    private val menuInputProcessor: MenuInputProcessor

    private val menuButtonRepository: MenuButtonRepository

    init {
        val menuComponent: MenuComponent = DaggerMenuComponent.builder()
            .applicationComponent((Gdx.app.applicationListener as CaveDroidApplication).applicationComponent)
            .build()

        menuRenderer = menuComponent.menuRenderer
        menuInputProcessor = menuComponent.menuInputProcessor
        menuButtonRepository = menuComponent.menuButtonRepository
    }

    fun resetMenu() {
        menuButtonRepository.setCurrentMenu("main")
    }

    override fun show() {
        resetMenu()
        Gdx.input.inputProcessor = menuInputProcessor
    }

    override fun render(delta: Float) {
        menuRenderer.render(delta)
    }

    override fun resize(width: Int, height: Int) {
        super.resize(width, height)
        menuRenderer.onResize()
    }

    override fun pause() {
    }

    override fun resume() {
    }

    override fun hide() {
    }

    override fun dispose() {
        menuRenderer.dispose()
        menuButtonRepository.dispose()
    }
}
